helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -190,42 +184,44 @@ private InputFields readInput(String content)
+        LOG.error(
+            "Instance " + instance + " in both ActivateInstances and DeactivateInstances list.");
+      }
+    }
+
+    // Add instances to current liveInstances
     ClusterState clusterState = new ClusterState();
     ConfigAccessor cfgAccessor = getConfigAccessor();
     HelixDataAccessor dataAccessor = getDataAccssor(clusterId);
     clusterState.resources = dataAccessor.getChildNames(dataAccessor.keyBuilder().idealStates());
     // Add existing live instances and new instances from user input to instances list.
-    clusterState.instances = dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances());
-    clusterState.instances.addAll(inputFields.newInstances);
-
-    // Check if to be removed instances and old instances in swap node exist in live instance.
-    if (!inputFields.nodeSwap.isEmpty() || !inputFields.instancesToRemove.isEmpty()) {
-      Set<String> liveInstanceSet = new HashSet<>(clusterState.instances);
-      for (Map.Entry<String, String> nodeSwapPair : inputFields.nodeSwap.entrySet()) {
-        if (!liveInstanceSet.contains(nodeSwapPair.getKey())) {
-          throw new InvalidParameterException("Invalid input: instance [" + nodeSwapPair.getKey()
-              + "] in SwapInstances does not exist in cluster.");
-        }
-      }
-      for (String instanceToRemove : inputFields.instancesToRemove) {
-        if (!liveInstanceSet.contains(instanceToRemove)) {
-          throw new InvalidParameterException("Invalid input: instance [" + instanceToRemove
-              + "] in RemoveInstances does not exist in cluster.");
-        }
-      }
-      if (!inputFields.instancesToRemove.isEmpty()) {
-        clusterState.instances.removeIf(inputFields.instancesToRemove::contains);
-      }
-    }
+    Set<String> liveInstancesSet =
+        new HashSet<>(dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances()));
+    liveInstancesSet.addAll(
+        inputFields.activatedInstances.stream().filter(s -> !liveInstancesSet.contains(s))
+            .collect(Collectors.toList()));
 
     // Read instance and cluster config.
-    // It will throw exception if there is no instanceConfig for newly added instance.
-    for (String instance : clusterState.instances) {
+    // Throw exception if there is no instanceConfig for activatedInstances instance.
+    for (String instance : liveInstancesSet) {
       InstanceConfig config = cfgAccessor.getInstanceConfig(clusterId, instance);
+      // override HELIX_ENABLED field.
+      if (inputFields.activatedInstances.contains(instance)) {
+        config.setInstanceEnabled(true);
+      }
+      if (inputFields.deactivatedInstances.contains(instance)) {
+        config.setInstanceEnabled(false);
+      }
       clusterState.instanceConfigs.add(config);
     }
-    clusterState.clusterConfig = cfgAccessor.getClusterConfig(clusterId);
+    liveInstancesSet.removeAll(inputFields.deactivatedInstances);
```

Comment:
nit, move to line 207 so all live instance operations happen at the same place?

Reply:
If we move this line to 207, then instances in deactivatedInstances list will be override to be in `out of cluster` state rather than `offline & disabled` state..?

Reply:
It is not a correct design to check instance config according to live instance. After you fix this, then the concern you described will go away.

Reply:
TFTR, updated